### ISSUE TITLE
Streaming non-u8 number parsers request too many bytes

### ISSUE THREAD

**Issue by Palladinium** [Geal/nom#1055]
- Rust version: `rustc 1.40.0-nightly (7979016af 2019-10-20)`
- nom version : `5.0.1`
- nom compilation features used: Default features

The primitive streaming parsers in `nom::number::streaming` return `Incomplete(Needed(n))`, where `n` is their full size in bytes, even in cases where fewer bytes are needed.

All `le_*`/`be_*` streaming parsers other than the `u8` and `i8` parsers are affected.

There are many doc tests to check this strange behaviour, so I'm not sure if this is for some reason intended.

## Test case

```
extern crate nom;

fn main() {
  assert_eq!(
    nom::number::streaming::be_i16(&[0x12]),
    Err(nom::Err::<()>::Incomplete(nom::Needed::Size(1)))
  );
}
```

**Comment by str4d** [comment]
This affects numerous other streaming parsers, such as `take`, `tag`, and the various line ending parsers. It would be great to know if this is intended behaviour or not, because at the outer level of a parser there is no context provided as to what fraction of `Needed(n)` is already in the buffer (and thus how much more to read from some `R: Read` I might be pulling data from).

**Comment by Geal** [comment]
that was a mistake, I'll do a pass on other parsers to make sure the behaviour is correct

**Comment by Geal** [comment]
So it's a bit tricky tondo that properly for tag and take when parsing a &str, since calculating the correct size would require counting chars. I'm tempted to just make them return `Needed::Unknown`, but maybe it could return the correct size for `&[u8]` where there's no additional cost

**Comment by Geal** [comment]
alright, this should be fixed with c5e0dc2, and it separates properly the `&[u8]` case where calculating the remaining number of byte is straightforward, and the `&str` case where we cannot really know

**Comment by Geal** [comment]
nom 6 is released now :partying_face:
